checker/src/org/checkerframework/checker/nullness/KeyForAnnotatedTypeFactory.java

```diff
@@ -469,68 +467,52 @@ public AnnotationMirror createKeyForAnnotationMirrorWithValue(String value) {
-      boolean unknownReceiver = false;
-
-      if (flowExprContext.receiver == null || flowExprContext.receiver.containsUnknown()) {
-          // If the receiver is unknown, we will try local variables
-
-          store = getStoreBefore(t);
-          unknownReceiver = true; // We could use store != null for this check, but this is clearer.
-      }
-
       if (anno != null) {
+          boolean unknownReceiver = flowExprContext.receiver == null || flowExprContext.receiver.containsUnknown();
           boolean valuesChanged = false; // Indicates that at least one value was changed in the list.
           LinkedHashSet<String> newValues = new LinkedHashSet<String>();
 
           List<String> values = AnnotationUtils.getElementValueArray(anno, "value", String.class, false);
           for (String s: values) {
-              boolean localVariableFound = false;
-
-              if (unknownReceiver) {
-                  // If the receiver is unknown, try a local variable
-                  CFAbstractValue<?> val = store.getValueOfLocalVariableByName(s);
-
-                  if (val != null) {
-                      newValues.add(s);
-                      // Don't set valuesChanged to true since local variable names are already canonicalized
-                      localVariableFound = true;
-                  }
+              Receiver varTypeReceiver = null;
+
+              try {
+                  varTypeReceiver = FlowExpressionParseUtil.parse(s, flowExprContext, path);
+              } catch (FlowExpressionParseException e) {
+            	  // Canonicalization is a best-effort approach and it should not cause an exception
+            	  // to be thrown if an expression cannot be parsed. If an expression that must be
+            	  // canonicalized cannot be because of an inability to parse it, a type checking
+            	  // error will be issued later if the expression is compared to its canonical equivalent.
+            	  // For example, if @KeyFor("#1") could not be canonicalized to @KeyFor("var1"), and
+            	  // it is later compared to a @KeyFor("var1") annotation that was written by the user,
+            	  // a type checking error will result.
               }
 
-              if (localVariableFound == false) {
+              if (unknownReceiver // The receiver type was unknown initially, and ...
+                      && (varTypeReceiver == null
+                      || varTypeReceiver.containsUnknown()) // ... the receiver type is still unknown after a call to parse
+                      ) {
+                  // parse did not find a static member field. Try a nonstatic field.
+
                   try {
-                      varTypeReceiver = FlowExpressionParseUtil.parse(s, flowExprContext, path);
+                      varTypeReceiver = FlowExpressionParseUtil.parse("this." + s, // Try a field in the current object. Do not modify s itself since it is used in the newValue.equals(s) check below.
```

Comment:
Fixed.
I'm not sure why the directory match was not a problem in the mainline. I see that Mike fixed it - we can ask him tomorrow.